source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc

```diff
@@ -0,0 +1,139 @@
+
+constexpr auto filter_metadata_key = "com.amazonaws.lambda";
+constexpr auto egress_gateway_metadata_key = "egress_gateway";
+
+void setHeaders(Http::HeaderMap& headers, absl::string_view function_name) {
+  headers.setMethod(Http::Headers::get().MethodValues.Post);
+  headers.setPath(fmt::format("/2015-03-31/functions/{}/invocations", function_name));
+  headers.setCopy(Http::LowerCaseString{"x-amz-invocation-type"}, "RequestResponse");
+}
+
+/**
+ * Determines if the target cluster has the AWS Lambda metadata on it.
+ */
+bool isTargetClusterLambdaGateway(Upstream::ClusterInfo const& cluster_info) {
+  using ProtobufWkt::Value;
+  const auto& filter_metadata_map = cluster_info.metadata().filter_metadata();
+  auto metadata_it = filter_metadata_map.find(filter_metadata_key);
+  if (metadata_it == filter_metadata_map.end()) {
+    return false;
+  }
+
+  auto egress_gateway_it = metadata_it->second.fields().find(egress_gateway_metadata_key);
+  if (egress_gateway_it == metadata_it->second.fields().end()) {
+    return false;
+  }
+
+  if (egress_gateway_it->second.kind_case() != Value::KindCase::kBoolValue) {
+    return false;
+  }
+
+  return egress_gateway_it->second.bool_value();
+}
+
+} // namespace
+
+FilterSettings::FilterSettings(const std::string& arn, bool payload_passthrough)
+    : arn_(arn), payload_passthrough_(payload_passthrough) {}
+
+Filter::Filter(const FilterSettings& settings,
+               const std::shared_ptr<Extensions::Common::Aws::Signer>& sigv4_signer)
+    : settings_(settings), sigv4_signer_(sigv4_signer) {}
+
+absl::optional<Arn> Filter::calculateRouteArn() {
+  if (!decoder_callbacks_->route() || !decoder_callbacks_->route()->routeEntry()) {
+    return absl::nullopt;
+  }
+  const auto* route_entry = decoder_callbacks_->route()->routeEntry();
+  const auto* settings = route_entry->mostSpecificPerFilterConfigTyped<FilterSettings>(
+      HttpFilterNames::get().AwsLambda);
+  if (!settings) {
+    return absl::nullopt;
+  }
+
+  return parse_arn(settings->arn());
+}
+
+Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool end_stream) {
+
+  auto route_arn = calculateRouteArn();
+  if (route_arn.has_value()) {
+    auto cluster_info_ptr = decoder_callbacks_->clusterInfo();
+    ASSERT(cluster_info_ptr);
+    if (!isTargetClusterLambdaGateway(*cluster_info_ptr)) {
+      skip_ = true;
+      return Http::FilterHeadersStatus::Continue;
+    }
+    arn_.swap(route_arn);
+  } else {
+    arn_ = parse_arn(settings_.arn());
+    if (!arn_.has_value()) {
+      ENVOY_LOG(error, "Unable to parse Lambda ARN {}.", settings_.arn());
+      skip_ = true;
+      return Http::FilterHeadersStatus::Continue;
+    }
+  }
+
+  if (end_stream) {
+    // We must NOT sign this header, because it causes signature miscalculation at Lambda's side.
+    // We _could_ remove the header, sign the request, and then tack it back on. But, I don't see
+    // value in that header when targeting AWS Lambda.
```

Comment:
These comments apply to L105-107 (not L104), right?

Reply:
fixed